joeynmt/model.py

```diff
@@ -383,6 +386,265 @@ def length_norm(l: int) -> float:
+                alive_seq_old = alive_seq
+                alive_seq = torch.full([0,step+1], bos_index, dtype=torch.long, device=device)
+                alive_score = torch.zeros([0], device=device)
+                for alive_index_old, batch_index in enumerate(range(batch_size)):
+
+                    if not are_all_beam_finished[batch_index]:
+                        thresholds[batch_index, step] = (beam_score[batch_index,-1] + runnerup_score[batch_index]) /2
+
+                        # find unfinished sequence
+                        if are_all_beam_finished_new[batch_index]:
+                            if not runnerup_finished[batch_index]:
+                                seq = runnerup_seq[batch_index]
+                            else:
+                                sorted_index = beam_vocab_score[batch_index].argsort(descending=True)  # (beam_size*trg_vocab_size)
+                                word_index = sorted_index.fmod(trg_vocab_size)
+                                unfinished = ~ word_index.eq(eos_index)
+                                first_unfinished_index = (unfinished * torch.arange(unfinished.shape[0],0,-1)).argmax()
+                                beam_origin_index = sorted_index[first_unfinished_index].floor_divide(trg_vocab_size)
+                                seq = torch.cat([
+                                    beam_seq_old[beam_origin_index],
+                                    word_index[first_unfinished_index].unsqueeze(-1)
+                                ])  # (step+1)
+                            alive_seq = torch.vstack([alive_seq, seq])
+
+                    else:
+                        # prep
+                        score, word_index = alive_vocab_score[alive_index_old].sort(descending=True)  # (trg_vocab_size)
+                        unfinished = ~ word_index.eq(eos_index)
+                        first_unfinished_index = (unfinished * torch.arange(unfinished.shape[0], 0, -1)).argmax()
+                        seq = torch.cat([
+                            alive_seq_old[alive_index_old],
+                            word_index[first_unfinished_index].unsqueeze(-1)
+                        ])  # (step+1)
+                        # comp threshold and alive_seq
+                        th_up = beam_score[batch_index,-1]
+                        try:
+                            th_dn = score[score<th_up].max()
+                        except RuntimeError:  # for the case that all scores are over the th_up
+                            th_dn = th_up - 0.1
+                        thresholds[batch_index, step] = (th_up + th_dn) /2
+                        alive_seq = torch.vstack([alive_seq, seq])
+
+                # backup beam_seq
+                beam_seq_of_all_steps[step] = beam_seq
+
+                # reshape `beam_seq` to its original size
+                beam_seq = beam_seq.reshape(batch_size * beam_size, step+1)  # (batch_size*beam_size, hyp_len)
```

Comment:
こちらで行っている `beam_seq` に対する更新処理が `else` の際には行われず、[こちら](https://github.com/ciela/reinforce-joey/pull/1/files#diff-fa49fcb36f3801488183d865b2745687110ade063fb975a1429c1c3501c0a28aR486)でサイズが合わずエラーになりますね。
`alive_seq` との兼ね合いもあると思いますのでどう対処すべきか考えますね。

```
    threasholds, _ = self._compute_threshold_by_vanilla_beam_search(
  File "/home/ciela/src/github.com/ciela/reinforce-joey/joeynmt/model.py", line 488, in _compute_threshold_by_vanilla_beam_search
    decoder_input = torch.vstack([beam_seq, alive_seq])  # (batch_size * beam_size + finished_batch_size, step)
RuntimeError: Tensors must have same number of dimensions: got 2 and 3
```

Reply:
あ、僕もそう思って移動してみたのですが、`alive_seq` の方とのサイズがまた合わなくなってしまって同じ箇所でエラー起こるんですよね。

```py3
log.debug(f'{beam_seq.size()=}, {alive_seq.size()=}')
decoder_input = torch.vstack([beam_seq, alive_seq])  # (batch_size * beam_size + finished_batch_size, step)
```

上記のようなログを仕込むと

```
[D 230207 08:41:42 model:486] beam_seq.size()=torch.Size([20, 1]), alive_seq.size()=torch.Size([0, 1])
[D 230207 08:41:42 model:486] beam_seq.size()=torch.Size([20, 2]), alive_seq.size()=torch.Size([0, 1])
```

こんな感じになってまして、テンソル2次元目のサイズがそれぞれで合わなくなってしまっていました。

`alive_seq` も同様に if 外に出そうと思ったのですが、このテンソルに追加されていっている `seq` が[こちらの `else`](https://github.com/ciela/reinforce-joey/pull/1/files#diff-fa49fcb36f3801488183d865b2745687110ade063fb975a1429c1c3501c0a28aR638-R640) の方では現れてないためどう作って上げるべきなのか相談したかった感じでした。

Reply:
`alive_seq` については else 内でこんな感じで横の長さ要素数を増やしてあげるとかですかね・・・？

```py
else:
    # calc threshold (Since this is the final step, there is no need for `alive_*` anymore.)
    thresholds[:, step] = beam_score[:,(n_best-1):(n_best+1)].mean(dim=-1)
    alive_seq = alive_seq.reshape(-1, alive_seq.size(1) + 1)
```

Reply:
いったんこちらで対応してみましたー（問題あれば後ほど取り消しますね）
https://github.com/ciela/reinforce-joey/pull/1/commits/2269faf7bfda54b3398cbac091282b9f4dd57cbf

Reply:
大元の587行目のif文の条件に間違いがありました。すみません。
` if step < max_output_length-1 and n_best < beam_size:`
ではなく、正しくは↓になります。
` if step < max_output_length-1 and n_best == beam_size:`

先程の修正にあわせて、こちらの修正も適用してみていただけますでしょうか

Reply:
https://github.com/ciela/reinforce-joey/commit/2269faf7bfda54b3398cbac091282b9f4dd57cbf の対応、すみません！

if文の条件を直せば、`else`に入るの可能性があるは最終ステップの場合のみ（`alive_seq` はもう必要ない）なので、予期せぬ挙動のときにエラー終了となるように、`else`内で`alive_seq`の更新処理は一旦消してしまって大丈夫です

Reply:
あ、なるほど！
承知しましたー、そしたら 2269faf の件は revert しておきますー

Reply:
こちらで諸々帳尻合わせましたー
https://github.com/ciela/reinforce-joey/compare/069d1b3..bc79b83

Reply:
すいません、上記の比較URL、修正いたしました。
別のコミット分（デバイス指定）も含まれていますが今回の件についても合わせて修正出来たと思いますー